### ISSUE TITLE
Sybase linq issue with expression builders

### ISSUE THREAD

**Issue by portikcs** [linq2db/linq2db#1324]
We have an issue, with the following type of call 
```cs
var query = from a in TableA
                   join b in TableB on a.Id equals b.Id
                   where a.Id > 1
                  select new {a.Name, b.Location};
```
If we run this query it is working fine, but if we run it after a call with extension method like this 
```cs
var query = TableA.Join(TableB, SqlJoinType.Inner,
                (a, b) => a.Id == b.Id,
                 (a,b) => new {a.Name, b.Location});
```
it is failing.
```     
Exception message: System.InvalidCastException: 'Specified cast is not valid.'
Stack trace: short version at LinqToDB.Linq.Builder.AllJoinsLinqBuilder.BuildMethodCall(ExpressionBuilder builder, MethodCallExpression methodCall, BuildInfo buildInfo)

LinqToDB.Linq.Builder.AllJoinsLinqBuilder line 50

Full stack: 
   at LinqToDB.Linq.Builder.AllJoinsLinqBuilder.BuildMethodCall(ExpressionBuilder builder, MethodCallExpression methodCall, BuildInfo buildInfo)

   at LinqToDB.Linq.Builder.MethodCallBuilder.BuildSequence(ExpressionBuilder builder, BuildInfo buildInfo)

   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildSequence(BuildInfo buildInfo)

   at LinqToDB.Linq.Builder.WhereBuilder.BuildMethodCall(ExpressionBuilder builder, MethodCallExpression methodCall, BuildInfo buildInfo)

   at LinqToDB.Linq.Builder.MethodCallBuilder.BuildSequence(ExpressionBuilder builder, BuildInfo buildInfo)

   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildSequence(BuildInfo buildInfo)

   at LinqToDB.Linq.Builder.SelectBuilder.BuildMethodCall(ExpressionBuilder builder, MethodCallExpression methodCall, BuildInfo buildInfo)

   at LinqToDB.Linq.Builder.MethodCallBuilder.BuildSequence(ExpressionBuilder builder, BuildInfo buildInfo)

   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildSequence(BuildInfo buildInfo)

   at LinqToDB.Linq.Builder.TakeSkipBuilder.BuildMethodCall(ExpressionBuilder builder, MethodCallExpression methodCall, BuildInfo buildInfo)

   at LinqToDB.Linq.Builder.MethodCallBuilder.BuildSequence(ExpressionBuilder builder, BuildInfo buildInfo)

   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildSequence(BuildInfo buildInfo)

   at LinqToDB.Linq.Builder.TakeSkipBuilder.BuildMethodCall(ExpressionBuilder builder, MethodCallExpression methodCall, BuildInfo buildInfo)

   at LinqToDB.Linq.Builder.MethodCallBuilder.BuildSequence(ExpressionBuilder builder, BuildInfo buildInfo)

   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildSequence(BuildInfo buildInfo)

   at LinqToDB.Linq.Builder.ExpressionBuilder.Build[T]()

   at LinqToDB.Linq.Query`1.CreateQuery(IDataContext dataContext, Expression expr)
```

### Steps to reproduce
Run the second query then the first one.

I have found a fix for the issue,  it is in class LinqToDB.Linq.Builder.AllJoinsLinqBuilder, method CanBuildMethodCall, the old code is like this :
```cs
methodCall.IsQueryable("Join") && methodCall.Arguments.Count == 5 ||
 methodCall.IsQueryable("InnerJoin", "LeftJoin", "RightJoin", "FullJoin") && methodCall.Arguments.Count == 4|| methodCall.IsQueryable("CrossJoin") && methodCall.Arguments.Count == 3;
```
The fix i to change it to :
```cs
(methodCall.IsQueryable("Join") && methodCall.Arguments.Count == 5 && methodCall.Method.DeclaringType == typeof(LinqExtensions))||
(methodCall.IsQueryable("InnerJoin", "LeftJoin", "RightJoin", "FullJoin") && methodCall.Arguments.Count == 4) ||(methodCall.IsQueryable("CrossJoin") && methodCall.Arguments.Count == 3);
```
Let me to explain:
I some of f the cases the AllJoinBuilder is chosen to build the expression for first type of query, what is wrong, so i blocked the AllJoinsLinqBuilder to be selected for first type  queries and in these cases ExpressionBuilder is  using the correct builder JoinBuilder. 

I did some testing and it was working fine after fix, but i do not have such a complex test cases like you, so the fix should be tested further.

I did not managed to create a branch and add the fix there, so please take in consideration and let me know.

### Environment details
linq2db version: 2.3.0
Database Server:  Sybase ASE Enterprise
Database Provider: *E.g. Npgsql 3.2.3*
Operating system: Windows Server the version is not provided
Framework version: .NET Framework 4.5.2

**Comment by sdanyliv** [comment]
@portikcs, thanks for reporting, will handle this.

**Comment by sdanyliv** [comment]
Looks like it is more detailed explanation of issue #1209

**Comment by sdanyliv** [comment]
Fixed in bb1ef44f231151195c2c0be8bd844606c0e8d542
